tests/internal/symbol_db/test_symbols.py

```diff
@@ -0,0 +1,203 @@
+from importlib.machinery import ModuleSpec
+from pathlib import Path
+from types import ModuleType
+
+import pytest
+
+from ddtrace.internal.symbol_db.symbols import Scope
+from ddtrace.internal.symbol_db.symbols import ScopeData
+from ddtrace.internal.symbol_db.symbols import ScopeType
+from ddtrace.internal.symbol_db.symbols import Symbol
+from ddtrace.internal.symbol_db.symbols import SymbolType
+
+
+def test_symbol_from_code():
+    def foo(a, b, c=None):
+        loc = 42
+        return loc
+
+    symbols = Symbol.from_code(foo.__code__)
+    assert {s.name for s in symbols if s.symbol_type == SymbolType.ARG} == {"a", "b", "c"}
+    assert {s.name for s in symbols if s.symbol_type == SymbolType.LOCAL} == {"loc"}
+
+
+def test_symbols_class():
+    class Sym:
+        def __init__(self):
+            self._foo = "foo"
+
+        @property
+        def foo(self):
+            return self._foo
+
+        @foo.setter
+        def _(self, value):
+            self._foo = value
+
+        @classmethod
+        def bar(cls):
+            pass
+
+        @staticmethod
+        def baz():
+            pass
+
+        def gen(n=10):
+            yield from range(n)
+
+        async def coro(b):
+            oroc = 42
+            yield oroc
+
+    module = ModuleType("test")
+    module.Sym = Sym
+    module.__spec__ = ModuleSpec("test", None)
+    module.__spec__.origin = __file__
+
+    scope = Scope.from_module(module)
+
+    (class_scope,) = scope.scopes
+    assert class_scope.name == "Sym"
+
+    (field,) = (s for s in class_scope.symbols if s.symbol_type == SymbolType.FIELD)
+    assert field.name == "_foo"
+
+    assert {s.name for s in class_scope.scopes if s.scope_type == ScopeType.FUNCTION} == {
+        "__init__",
+        "bar",
+        "baz",
+        "coro",
+        "foo",
+        "gen",
+    }
+
+
+def test_symbols_decorators():
+    """Test that we get the undecorated functions from a module scope."""
```

Comment:
Good point. I'm still not quite sure what Python-specific features are worth including in the payloads. Decorators are not generally easy to retrieve, as they too are syntactic sugar (generally they are callable objects that take a function and return a possibly new function), so we don't have an easy way of retrieving them. We do have annotations in Python though, but they are primarily used for type hints. We could perhaps extract that information when available, but because type hints are not enforced, the backend/UI won't be able to rely on their presence all the time.

Reply:
Type hints would be great! If they exist, they could be used for auto-complete in the expression language :)

Reply:
Actually, type hints should be added to `language_specifics.return_type` for functions and for symbols just in the symbol itself as `type`.